### ISSUE TITLE
Hot module reloading doesn't work

### ISSUE THREAD

**Issue by amcsi** [kriasoft/react-starter-kit#932]
I have cloned latest master which is 048e67f, did `npm install && npm start`, and hot reloading doesn't work.

The browser window was opened with localhost:3001, and changes saved in my editor does trigger recompiling and the changes can be seen when refreshing, but there's no hot module replacement (or even auto refreshing) happening.

**Comment by amcsi** [comment]
The issue seems to still be there in 82a54939ee3407ba8cb90cc8111366be01813081 which is before the react-hot-loader v3 commit  (2f7d127894fd547700a784c57f83af276b3e24bd )

**Comment by frenzzy** [comment]
Please read https://github.com/kriasoft/react-starter-kit/issues/907#issuecomment-252641016 or provide more info what exactly doesn't work (which component, browser's console logs).

**Comment by amcsi** [comment]
It's happening in the Header and Footer for example

Console logs after making a change:

```
[HMR] connected
client.js:128 [HMR] bundle rebuilding
client.js:132 [HMR] bundle rebuilt in 4266ms
process-update.js:27 [HMR] Checking for updates on the server...
process-update.js:100[HMR] Updated modules:
process-update.js:102[HMR]  - ./src/components/Footer/Footer.js
process-update.js:102[HMR]  - ./src/components/Layout/Layout.js
process-update.js:102[HMR]  - ./src/routes/home/Home.js
process-update.js:102[HMR]  - ./src/routes/contact/Contact.js
process-update.js:102[HMR]  - ./src/routes/login/Login.js
process-update.js:102[HMR]  - ./src/routes/register/Register.js
process-update.js:102[HMR]  - ./src/routes/content/Content.js
process-update.js:102[HMR]  - ./src/routes/notFound/NotFound.js
process-update.js:102[HMR]  - ./src/routes/notFound/index.js
process-update.js:102[HMR]  - ./src/routes/index.js
process-update.js:102[HMR]  - ./src/routes/content/index.js
process-update.js:102[HMR]  - ./src/routes/register/index.js
process-update.js:102[HMR]  - ./src/routes/login/index.js
process-update.js:102 [HMR]  - ./src/routes/contact/index.js
process-update.js:102 [HMR]  - ./src/routes/home/index.js
process-update.js:107 [HMR] App is up to date.
```

A sample of the output of `npm start`:

```
"C:\f\PhpStorm 2016.1\bin\runnerw.exe" C:\p\nodejs\node.exe C:\p\nodejs\node_modules\npm\bin\npm-cli.js run-script start

> @ start C:\0\react-starter-kit
> babel-node tools/run start

[16:15:56] Starting 'start'...
[16:15:56] Starting 'clean'...
[16:15:56] Finished 'clean' after 113 ms
[16:15:56] Starting 'bound copy'...
[16:15:56] Finished 'bound copy' after 171 ms
webpack built 3c7d715a907917574216 in 13566ms
Child
    Time: 13566ms
                                  Asset     Size  Chunks             Chunk Names
           main.js?3c7d715a907917574216  1.63 MB       0  [emitted]  main
        admin.1.js?3c7d715a907917574216  6.17 kB       1  [emitted]  admin
       main.js.map?3c7d715a907917574216  1.93 MB       0  [emitted]  main
    admin.1.js.map?3c7d715a907917574216   7.7 kB       1  [emitted]  admin
Child
    Time: 14211ms
                  Asset    Size  Chunks             Chunk Names
        ../../server.js  213 kB       0  [emitted]  main
    ../../server.js.map  235 kB       0  [emitted]  main
webpack: bundle is now VALID.
[16:16:17] The server is running at http://localhost:3000/
[BS] Proxying: http://localhost:3000
[BS] Access URLs:
 -------------------------------------
       Local: http://localhost:3001
    External: http://192.168.0.17:3001
 -------------------------------------
          UI: http://localhost:3002
 UI External: http://192.168.0.17:3002
 -------------------------------------
[BS] Watching files...
[16:16:20] Finished 'start' after 23987 ms
webpack: bundle is now INVALID.
webpack building...
webpack built 2fa648e7a7867504bee2 in 3873ms
Child
    Time: 3873ms
                                       Asset      Size  Chunks             Chunk Names
                main.js?2fa648e7a7867504bee2   1.63 MB       0  [emitted]  main
             admin.1.js?2fa648e7a7867504bee2   6.11 kB       1  [emitted]  admin
        0.3c7d715a907917574216.hot-update.js    2.5 kB       0  [emitted]  main
        3c7d715a907917574216.hot-update.json  36 bytes          [emitted]  
            main.js.map?2fa648e7a7867504bee2   1.93 MB       0  [emitted]  main
    0.3c7d715a907917574216.hot-update.js.map   2.02 kB       0  [emitted]  main
Child
    Time: 3432ms
                  Asset    Size  Chunks             Chunk Names
        ../../server.js  213 kB       0  [emitted]  main
    ../../server.js.map  235 kB       0  [emitted]  main
webpack: bundle is now VALID.
[16:16:33] The server is running at http://localhost:3000/
webpack: bundle is now INVALID.
webpack building...
webpack built 70d27a24fb31b6274ff9 in 4884ms
Child
    Time: 4884ms
                                       Asset      Size  Chunks             Chunk Names
                main.js?70d27a24fb31b6274ff9   1.63 MB       0  [emitted]  main
        0.2fa648e7a7867504bee2.hot-update.js   2.66 kB       0  [emitted]  main
        2fa648e7a7867504bee2.hot-update.json  36 bytes          [emitted]  
            main.js.map?70d27a24fb31b6274ff9   1.93 MB       0  [emitted]  main
    0.2fa648e7a7867504bee2.hot-update.js.map   2.13 kB       0  [emitted]  main
Child
    Time: 4955ms
                  Asset    Size  Chunks             Chunk Names
        ../../server.js  213 kB       0  [emitted]  main
    ../../server.js.map  235 kB       0  [emitted]  main
webpack: bundle is now VALID.
[17:27:13] The server is running at http://localhost:3000/
webpack: bundle is now INVALID.
webpack building...
webpack built 1f16479cf7431648de89 in 4266ms
Child
    Time: 4266ms
                                       Asset      Size  Chunks             Chunk Names
                main.js?1f16479cf7431648de89   1.63 MB       0  [emitted]  main
        0.70d27a24fb31b6274ff9.hot-update.js   2.66 kB       0  [emitted]  main
        70d27a24fb31b6274ff9.hot-update.json  36 bytes          [emitted]  
            main.js.map?1f16479cf7431648de89   1.93 MB       0  [emitted]  main
    0.70d27a24fb31b6274ff9.hot-update.js.map   2.13 kB       0  [emitted]  main
Child
    Time: 4340ms
                  Asset    Size  Chunks             Chunk Names
        ../../server.js  213 kB       0  [emitted]  main
    ../../server.js.map  235 kB       0  [emitted]  main
webpack: bundle is now VALID.
[17:28:35] The server is running at http://localhost:3000/

```

**Comment by frenzzy** [comment]
From https://github.com/kriasoft/react-starter-kit/issues/871#issuecomment-250741019

> After editing Home.js, both client-side and server-side bundles are being recompiled, the problem is that the client-side portion is reloaded faster, but there is an Ajax request hanging in pending state waiting for the server to respond which never does. Ideas how to solve this issue are welcome! :)

**Comment by amcsi** [comment]
I see. But isn't everyone getting this issue?

I'm using Windows 10 ny the way. Not sure if that matters.

**Comment by frenzzy** [comment]
Have you tried to check the hot updates on any other pages except the main one? For example hmr on `/login` page with any component works fine for me.

Have you use [`PureComponent`](https://facebook.github.io/react/docs/react-api.html#react.purecomponent) or [`shouldComponentUpdate`](https://facebook.github.io/react/docs/react-component.html#shouldcomponentupdate) anywhere in hierarchy of components? If so, you additionally need to [force update](https://facebook.github.io/react/docs/react-component.html#forceupdate) all components after hot update ([example](https://gist.github.com/frenzzy/3c09b527c587f0b2dd232195bc591bc7/revisions)).

**Comment by amcsi** [comment]
I'm doing all of this on the starter kit as it is so I'm not using anything that's not there.

I'm going to have to come back to you on this, but so far I've tried on 4 different components and none of them updated.

**Comment by frenzzy** [comment]
Do you see pending request on devtools network tab after hot changes?

![image](https://cloud.githubusercontent.com/assets/640669/19636527/879f9b70-99d2-11e6-957c-4222c6ddf5d4.png)

**Comment by amcsi** [comment]
@frenzzy yes I am indeed getting that same thing

And now I was on Ubuntu 16:04 and 048e67f

**Comment by petemill** [comment]
Also seeing same issue from latest source. No module refresh after browser receives module updates. Also see pending `graphql` request.

**Comment by karna41317** [comment]
still have the same problem

**Comment by frenzzy** [comment]
@karna41317 can you provide steps to reproduce?

**Comment by socrat3z** [comment]
Same problem on my windows 10 pc.
node: 7.5

Hot reloading stops working after merging react-intl branch.

**Comment by karna41317** [comment]
@frenzzy 

Run Redux branch locally in release or debug mode,  (BS will open default browser)
Refresh the window after you make any css or Js changes
click between different navigations Contat <->Login <-->about , At this point your browser will have loading indicator 
in console, network tab -> graphql request is in pending status

May be this is something to do with Chrome browser and socket connection.Not sure

**Comment by maksimkurb** [comment]
@socrat3z @frenzzy 
For me HMR started working on `react-intl` branch, when I changed [this part of client.js](https://github.com/kriasoft/react-starter-kit/blob/39a8ce6449b99b42dd44f5da2c6807f9612bfa60/src/client.js#L201-L208) to
```diff
// Enable Hot Module Replacement (HMR)
if (module.hot) {
  module.hot.accept('./router', () => {
    if (appInstance && appInstance.updater.isMounted(appInstance)) {
      // Force-update the whole tree, including components that refuse to update
      deepForceUpdate(appInstance);
    }
+    onLocationChange(currentLocation); // <<<< added onLocationChange() here, it was removed previously
  });
}
```

**Comment by langpavel** [comment]
Fixed in 6910350a295c08f4039e8ed942be82c4fc45d6df
Thanks @maksimkurb
